Sources/SwiftDocC/Infrastructure/ContentCache.swift

```diff
@@ -0,0 +1,93 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+extension DocumentationContext {
+    /// A cache for symbol and page content.
+    struct ContentCache<Value> {
+        private var storage = [ResolvedTopicReference: Value]()
+        private var symbolIndex = [String: ResolvedTopicReference]()
+        
+        /// Accesses the value for a given reference.
+        /// - Parameter reference: The reference to find in the cache.
+        subscript(reference: ResolvedTopicReference) -> Value? {
+            // Avoid copying the values if possible
+            _read { yield storage[reference] }
+            _modify { yield &storage[reference] }
+        }
+        
+        /// Adds a value to the cache for a given reference _and_ symbol ID.
+        /// - Parameters:
+        ///   - value: The value to add to the cache.
+        ///   - reference: The reference associated with that value.
+        ///   - symbolID: The symbol ID associated with that value.
+        mutating func add(value: Value, reference: ResolvedTopicReference, symbolID: String) {
+            symbolIndex[symbolID] = reference
+            storage[reference] = value
+        }
+        
+        /// Accesses the reference for a given symbol ID.
+        /// - Parameter symbolID: The symbol ID to find in the cache.
+        func reference(symbolID: String) -> ResolvedTopicReference? {
+            symbolIndex[symbolID]
+        }
+        
+        /// Accesses the value for a given symbol ID.
+        /// - Parameter symbolID: The symbol ID to find in the cache.
+        subscript(symbolID: String) -> Value? {
+            // Avoid copying the values if possible
+            _read { yield symbolIndex[symbolID].map { storage[$0]! } }
```

Comment:
Since it's not safe to add values to the cache from multiple threads and the only way to add a value to the `symbolIndex` is calling `add(value:reference:symbolID:)` which also adds the corresponding reference to the `storage`, it should be safe to assume that the reference exists in the storage.

Reply:
I documented this in https://github.com/apple/swift-docc/pull/806/commits/7322841116333dfb895746d347a31fd213eb147e